nncf/tensor/functions/tf_numeric.py

```diff
@@ -0,0 +1,534 @@
+# Copyright (c) 2024 Intel Corporation
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#      http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Callable, List, Optional, Tuple, Union
+
+import numpy as np
+import tensorflow as tf
+
+from nncf.tensor import TensorDataType
+from nncf.tensor import TensorDeviceType
+from nncf.tensor.definitions import TensorBackend
+from nncf.tensor.definitions import TypeInfo
+from nncf.tensor.functions import numeric as numeric
+
+DTYPE_MAP = {
+    TensorDataType.float16: tf.float16,
+    TensorDataType.bfloat16: tf.bfloat16,
+    TensorDataType.float32: tf.float32,
+    TensorDataType.float64: tf.float64,
+    TensorDataType.int8: tf.int8,
+    TensorDataType.int32: tf.int32,
+    TensorDataType.int64: tf.int64,
+    TensorDataType.uint8: tf.uint8,
+}
+
+DEVICE_MAP = {TensorDeviceType.CPU: "CPU", TensorDeviceType.GPU: "GPU"}
+
+DTYPE_MAP_REV = {v: k for k, v in DTYPE_MAP.items()}
+DEVICE_MAP_REV = {v: k for k, v in DEVICE_MAP.items()}
+
+
+@numeric.device.register(tf.Tensor)
+def _(a: tf.Tensor) -> TensorDeviceType:
+    return DEVICE_MAP_REV[a.device.split("/")[-1].split(":")[1]]
+
+
+@numeric.backend.register(tf.Tensor)
+def _(a: tf.Tensor) -> TensorBackend:
+    return TensorBackend.tf
+
+
+@numeric.squeeze.register(tf.Tensor)
+def _(a: tf.Tensor, axis: Optional[Union[int, Tuple[int, ...]]] = None) -> tf.Tensor:
+    with tf.device(a.device):
+        if axis is None:
+            return tf.squeeze(a)
+        if isinstance(axis, Tuple) and any(a.shape[i] != 1 for i in axis):
+            raise ValueError("Cannot select an axis to squeeze out which has size not equal to one")
+        return tf.squeeze(a, axis)
+
+
+@numeric.flatten.register(tf.Tensor)
+def _(a: tf.Tensor) -> tf.Tensor:
+    with tf.device(a.device):
+        return tf.reshape(a, [-1])
+
+
+@numeric.max.register(tf.Tensor)
+def _(a: tf.Tensor, axis: Optional[Union[int, Tuple[int, ...]]] = None, keepdim: bool = False) -> tf.Tensor:
+    with tf.device(a.device):
+        if axis is None:
+            return tf.reduce_max(a)
+        return tf.reduce_max(a, axis=axis, keepdims=keepdim)
```

Comment:
I've recognized that torch part have the same issue (`keepdims` is always `False` in case if `axis` is `None`). It works because `torch.allclose` function does not compare shapes (as well as in TensorFlow). Was it intentional? And should I fix torch part here as well if it was not?